specs/mpt-proof.md

```diff
@@ -0,0 +1,500 @@
+- to check whether `s_advices/c_advices` (at `is_modified` position)
+  present the hash of the next element in a proof
+
+Hash lookup looks like (for example for branch S):
+
+```
+lookup(S branch RLC, S branch length, s_advices RLC at the is_modified position in parent branch)
+```
+
+TODO: instead of 32 columns for `*_advices`, use only the RLC of `*_advices`.
+To integrate `*_advices` RLC into the computation of the whole branch RLC, we
+just need to compute `mult * *_advices_RLC` and add this to the current RLC value.
+
+The layout then be simply:
+`s_rlp1, s_rlp2, s_child_rlc, c_rlp1, c_rlp2, c_child_rlc`
+
+#### Key RLC in branch
+
+There are further columns (not shown in picture or table):
+
+- `node_index`
+- `is_modified`
+- `modified_node`
+
+In the example branch from the picture, `modified_node` would be 11.
+`node_index` is simply an index running from 0 to 15. `modified_node` is 1 at
+position 11, otherwise 0.
+
+Let's see a branch layout where we replace 32 columns of `s_advices` with
+one column `s_child_rlc` and 32 columns of `c_advices` with
+one column `c_child_rlc` (this simplification will most likely be implemented):
+
+```
+s_rlp1_0, s_rlp2_0, s_child0_rlc, c_rlp1_0, c_rlp2_0, c_child0_rlc
+s_rlp1_1, s_rlp2_1, s_child1_rlc, c_rlp1_1, c_rlp2_1, c_child1_rlc
+...
+s_rlp1_15, s_rlp2_15, s_child15_rlc, c_rlp1_15, c_rlp2_15, c_child15_rlc
+```
+
+With the three `modified_node` related columns:
+
+```
+s_rlp1_0, s_rlp2_0, s_child0_rlc, c_rlp1_0, c_rlp2_0, c_child0_rlc, 0 (node_index), 0 (is_modified), 11 (modified_node)
+s_rlp1_1, s_rlp2_1, s_child1_rlc, c_rlp1_1, c_rlp2_1, c_child1_rlc, 1 (node_index), 0 (is_modified), 11 (modified_node)
+...
+s_rlp1_11, s_rlp2_11, s_child11_rlc, c_rlp1_11, c_rlp2_11, c_child11_rlc, 11 (node_index), 1 (is_modified), 11 (modified_node)
+...
+s_rlp1_15, s_rlp2_15, s_child15_rlc, c_rlp1_15, c_rlp2_15, c_child15_rlc, 15 (node_index), 0 (is_modified), 11 (modified_node)
+```
+
+While navigating through branches, key RLC is being computed in the last branch child
+row. Let us say there are six branches with the following `modified_node` values:
+3, 11, 15, 14, 2, 1.
+
+These values present the first six nibbles of the key (after it is hashed).
+That means the first three bytes of the key are: `3 * 16 + 11`, `15 * 16 + 14`,
+`2 * 16 + 1`.
+
+To compute (partial) key RLC, we need to compute:
+
+```
+(3 * 16 + 11) + (15 * 16 + 14) * r + (2 * 16 + 1) * r^2
+```
+
+Once we reach the leaf node, we take the rest of the nibbles there and compute
+the full key RLC. This value needs to correspond to `key1` RLC.
+
+In the last branch child row, we compute `11 * 16 * key_rlc_mult`
+or `11 * key_rlc_mult` and add this value to the value in `key_rlc` row.
+
+Columns (not shown in picture or table):
+
+- `key_rlc_mult`
+- `key_rlc`
+- `sel1`
+- `sel2`
+
+Whether the factor 16 is used or not is determined by `sel1/sel2` columns.
+`sel1/sel2` are boolean values and it holds `sel1 + sel2 = 0`.
```

Comment:
I guess `sel1 + sel2 = 0` would indicate both `sel1` and `sel2` are 0. Should it be `sel1 + sel2 = 1`?

Reply:
Sorry @noel2004 , I haven't seen the question until now. But you are totally right, I will fix it!